trestle/core/catalog/catalog_reader.py

```diff
@@ -120,30 +125,96 @@ def read_catalog_from_markdown(self, md_path: pathlib.Path, set_parameters_flag:
+        imp_req_map: Dict[str, comp.ImplementedRequirement] = {}
         for group_path in CatalogInterface._get_group_ids_and_dirs(md_path).values():
             for control_file in group_path.glob('*.md'):
-                sort_id, imp_req = ControlReader.read_implemented_requirement(control_file, avail_comps, context)
+                sort_id, imp_req = ControlReader.read_implemented_requirement(control_file, context)
                 imp_req_map[sort_id] = imp_req
         return [imp_req_map[key] for key in sorted(imp_req_map.keys())]
 
+    @staticmethod
+    def _get_imp_req_for_control(ssp: ossp.SystemSecurityPlan, control_id: str) -> ossp.ImplementedRequirement:
+        for imp_req in as_list(ssp.control_implementation.implemented_requirements):
+            if imp_req.control_id == control_id:
+                return imp_req
+        imp_req = gens.generate_sample_model(ossp.ImplementedRequirement)
+        imp_req.control_id = control_id
+        ssp.control_implementation.implemented_requirements = as_list(
+            ssp.control_implementation.implemented_requirements
+        )
+        ssp.control_implementation.implemented_requirements.append(imp_req)
+        return imp_req
+
+    @staticmethod
+    def _get_imp_req_for_statement(
+        ssp: ossp.SystemSecurityPlan, control_id: str, statement_id: str
+    ) -> ossp.ImplementedRequirement:
+        control_imp_req: Optional[ossp.ImplementedRequirement] = None
+        for imp_req in as_list(ssp.control_implementation.implemented_requirements):
+            if imp_req.control_id == control_id:
+                control_imp_req = imp_req
+                if statement_id in [stat.statement_id for stat in as_list(imp_req.statements)]:
+                    return imp_req
+        # we didn't find imp_req with statement so need to make statement and/or imp_req
+        if not control_imp_req:
+            control_imp_req = gens.generate_sample_model(ossp.ImplementedRequirement)
+            control_imp_req.control_id = control_id
+            control_imp_req.statements = None
+            ssp.control_implementation.implemented_requirements = as_list(
+                ssp.control_implementation.implemented_requirements
+            )
+            ssp.control_implementation.implemented_requirements.append(control_imp_req)
+        statement = gens.generate_sample_model(ossp.Statement)
+        statement.statement_id = statement_id
+        statement.by_components = None
+        control_imp_req.statements = as_list(control_imp_req.statements)
+        control_imp_req.statements.append(statement)
+        return control_imp_req
+
+    @staticmethod
+    def _get_by_comp_from_imp_req(
+        imp_req: ossp.ImplementedRequirement, statement_id: str, comp_uuid: str
+    ) -> ossp.ByComponent:
+        if statement_id:
+            for statement in as_list(imp_req.statements):
+                if statement.statement_id == statement_id:
```

Comment:
A great use case for `as_filtered_list`

Reply:
This is another situation where the filtering is in a loop and it triggers a sonar warning - as discussed.  Until we have a workaround that avoids that warning (which can lead to an actual error when the code is pushed) I won't be able to use the tools that rely on a lambda for filtering - in a loop like this.